### ISSUE TITLE
Setting hugging face datasets in SpeechLM2

### ISSUE THREAD

**Issue by mhenrichsen** [NVIDIA-NeMo/NeMo#14268]
Hi,

I looked through your [Canary-qwen example config](https://github.com/NVIDIA/NeMo/blob/main/examples/speechlm2/conf/salm.yaml) and was wondering if it's possible to use HF datasets. If not, what format do the trainer expect the datasets to be in? 

Column names etc.

**Comment by pzelasko** [comment]
I'm in progress of writing a detailed tutorial about setting all this up, so in the meantime:

## HF datasets are not directly supported yet but I hope enabling this should be straightforward.
Our dataloaders are based on Lhotse, which already supports wrapping HF dataset:
https://github.com/lhotse-speech/lhotse/blob/fda1a986e5e1e72a14c82049b4ee709fc09a81e6/lhotse/cut/set.py#L2643-L2688
We'll only need to make sure there's a way to shuffle the underlying HF dataset with different seeds for different gpu+dataloader workers. If anybody is interested in picking this up before I can, the proper way would be to add a new `@data_type_parser` here:
https://github.com/NVIDIA/NeMo/blob/c17a0bfe5d71b4955750eadf671a65303203dee2/nemo/collections/common/data/lhotse/cutset.py#L100-L126

## What data formats are supported and how to specify them 
95% of dataloader code in NeMo is currently shared between all speech collections (asr, tts, enhancement, speechlm2, etc), and our approach has been to try and support as many formats as possible for everything, because people tend to have existing data in some specific format and requiring a copy is often prohibitive/throws them off. 
For speechlm2 specifically, [every format is being mapped into an iterable of `NeMoMultimodalConversation`](https://github.com/NVIDIA/NeMo/blob/c17a0bfe5d71b4955750eadf671a65303203dee2/nemo/collections/common/data/lhotse/cutset.py#L464-L505).

Generally supported formats:
* NeMo ASR manifest (tarred and non-tarred): https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/asr/datasets.html#preparing-custom-asr-data
* Lhotse CutSet manifests (normal and Lhotse Shar, ie tarred): https://lhotse.readthedocs.io/en/latest/cuts.html#cuts
* JSONL manifests with `NeMoMultimodalConversation` objects: https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/asr/datasets.html#configuring-multimodal-dataloading
 
This unit test setup is a great example how to format data directly as conversations:
https://github.com/NVIDIA/NeMo/blob/c17a0bfe5d71b4955750eadf671a65303203dee2/tests/collections/common/test_lhotse_multimodal_dataloading.py#L46-L92

Setting up the dataloader via config:
* Docs https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/asr/datasets.html#extended-multi-dataset-configuration-format
* Tutorial (look for NeMoMultimodalConversation specifically) https://github.com/NVIDIA/NeMo/blob/main/tutorials/multimodal/Multimodal%20Lhotse%20Dataloading.ipynb

The tutorial I mentioned will show end-to-end example of how to put together all this stuff, but for now just a bunch of examples to make this more digestible, using the SALM config you pointed to:

### Example config for dataloading from NeMo ASR manifests

```yaml
data:
  train_ds:
    ...
    input_cfg:
      - type: lhotse_as_conversation  # converts nemo and lhotse data to multimodal conversations 
        # non-tarred
        manifest_filepath: /path/to/manifest.json
        # tarred
        # manifest_filepath: /path/to/tarred/sharded_manifests/manifest__OP_0..127_CL_.json
        # tarred_audio_filepaths: /path/to/tarred/audio__OP_0..127_CL_.tar
        audio_locator_tag: ${model.audio_locator_tag}
        tags:
          # Uncomment below line to include a system prompt, if supported by the model's prompt formatter.
          # system_prompt: "some system prompt"
          context: "Transcribe the following:"
```

### Example config for dataloading from Lhotse manifests

```yaml
data:
  train_ds:
    ...
    input_cfg:
      - type: lhotse_as_conversation  # converts nemo and lhotse data to multimodal conversations 
        # non-tarred
        cuts_path: /path/to/cuts.jsonl.gz
        # tarred
        # shar_path: /path/to/lhotse_shar_data/
        audio_locator_tag: ${model.audio_locator_tag}
        tags:
          # Uncomment below line to include a system prompt, if supported by the model's prompt formatter.
          # system_prompt: "some system prompt"
          context: "Transcribe the following:"
```

### Example config for reading NeMoMultimodalConversation directly
```yaml
data:
  train_ds:
    ...
    input_cfg:
      - type: multimodal_conversation
        manifest_filepath: /path/to/conversation.jsonl
        # tarred
        # manifest_filepath: /path/to/tarred_conversations/sharded_manifests/manifest__OP_0..127_CL_.json
        # tarred_audio_filepaths: /path/to/tarred_conversations/audio__OP_0..127_CL_.tar
        audio_locator_tag: ${model.audio_locator_tag}
        # tags shouldn't be needed, your data will likely already include all prompts
```

**Comment by JRD971000** [comment]
Thanks @pzelasko; hi @mhenrichsen , is your issue resolved?

**Comment by Respaired** [comment]
@pzelasko 
@JRD971000 

May I ask you guys to show me an example of how does the json manifest looks like for training Canary Qwen ? I looked at the HF Repo, but I'm not sure I got it right.

btw, does it make any sense to use the tiny 180M Canary for the encoder?
